softwareComponents/isoreconfig/src/isomorphic.cpp

```diff
@@ -0,0 +1,132 @@
+#include <isoreconfig/isomorphic.hpp>
+#include <cassert>
+
+namespace rofi::isoreconfig {
+
+using namespace rofi::configuration;
+
+Point posToPoint( const Matrix& position )
+{
+    return Point{ position.at(0,3), position.at(1,3), position.at(2,3) };
+}
+
+Matrix pointToPos( const Point& point )
+{
+    Matrix result( arma::fill::eye );
+    for ( int i = 0; i < 3; ++i )
+        result(i,3) = point(i);
+    return result;
+}
+
+// Assumes <mod> is a UniversalModule
+// Returns pair of CoP (fst is module points, second is connection points)
+Points decomposeUniversalModule( 
+    const Module& mod, const Matrix& modPos )
+{
+    assert( mod.type == ModuleType::Universal );
+
+    std::span<const Component> comps = mod.components();
+    Points result;
+
+    for ( int compIndex = 0; compIndex < 6; ++compIndex )
+    {
+        // First six components are Roficoms
+        assert( comps[ compIndex ].type == ComponentType::Roficom );
+
+        Matrix relCompPos = modPos * mod.getComponentRelativePosition( compIndex );
+        result.push_back( posToPoint( relCompPos * matrices::translate( { -0.5, 0, 0 } ) ) );
+    } 
+
+    return result;
+}
+
+// Assumes the rofibot has been prepared
+// Assumes the rofibot consists of only UniversalModules
+std::array< Points, 2 > decomposeRofibot( const Rofibot& rb )
+{
+    auto [ ok, err ] = rb.isValid();
+    assert( ok );
+
+    std::array< Points, 2 > result;
+
+    for ( const Rofibot::ModuleInfo& modInf : rb.modules() )
+    {
+        std::array< Points, 2 > moduleClouds = decomposeUniversalModule( *modInf.module, *modInf.absPosition, modInf );
+
+        for ( const Point& p : moduleClouds[0] )
+            result[0].push_back( p );       
+    }
+
+    for ( const RoficomJoint& connection : rb.roficomConnections() )
+    {
+        Matrix pos = connection.getSourceModule( rb ).components()[connection.sourceConnector].getPosition();
+        result[1].push_back( posToPoint( pos * matrices::translate( { -0.5, 0, 0 } ) ) );
+    }
+
+    return result;
+}
+
+// Center of gravity does not take roficom connections into account
+Point getCenter( const Rofibot& rb )
+{
+    return getCenter( decomposeRofibot( rb )[0] );
+}
+
+Point getCenter( const Points& points )
+{
+    assert( points.size() >= 1 );
+    Point result = { 0, 0, 0 };
+    
+    for ( const Point& point : points )
+        result += point;
+
+    double pointCount = points.size();
+    return result / Point( { pointCount, pointCount, pointCount } );
+}
+
+double cubeNorm( const Point& vec )
```

Comment:
Tohle mi sem úplně nesedí, více bych dekomponoval co jsou pomocné operace nad vektory/maticemi, a co jsou operace nad rofibotem

Reply:
Máš tím na mysli, že bych měl tyto funkce rozdělit na skupiny v rámci jednoho souboru, nebo vytvořit více .cpp a případně i .hpp souborů, do kterých implementaci rozdělím?